### DISCUSSION TITLE
Global dirty state

### DISCUSSION THREAD

**Discussion by ju5t** [livewire/livewire#3537] [1. Help]
Hi there. We have a few full page components. On those pages we have actions/button that can either be destructive or where it's really helpful to have the most recent data. I know autosaving is a possible solution, but that's not for us.

To make things more user friendly I'd like to be able to check if there has been a change. This would allow us to warn users before running an action.

I don't see a way to check how many elements are dirty. I can see which elements are marked as 'to be checked', but that's it. Has anyone done this before?

**Comment by joshhanley** [comment]
@ju5t can you expand on it a bit more of what you mean?

**Reply by ju5t** [reply]
Our exact use case is this: let's say we have a model for a quote that has 5 fields. We have a component that is used to update an existing quote. All 5 columns are wire:model'ed (new word?) on this component. The component has a save and a print button.

If the user changes any of these 5 fields and they click on the print button, their changes are lost.

A user friendly solution to this is to show a warning when they hit the print button to show that there are unsaved changes. I thought it would be best described as a global dirty state (as `wire:dirty` only targets an element, I think?). In the end, all I like is to check if there are any unsaved changes. Perhaps I thought of the wrong name for it :)

**Comment by joshhanley** [comment]
@ju5t right, yep that makes sense!

If you are using model binding, Laravel has an `->isDirty()` method on a model which you can use for that.

Otherwise you would need to write a manual check, where you get the current record from the DB and compare the properties in the component.

Hope this helps!

**Reply by ju5t** [reply]
@joshhanley oh, `->isDirty()`! How can I forget :)

Unfortunately though, not everything is or can be wired to a model. I was kind of hoping that I could mark all of our input fields with `wire:dirty` and have some global state in Livewire (preferably in the JS object as it saves a round trip), to count how many fields are marked as dirty. I guess that's not an option :(

**Reply by joshhanley** [reply]
@ju5t yeah fair enough. Not sure if you can hook into `wire:dirty` or not. I guess you could output a specific class name using `wire:dirty` and do a count maybe using Alpine for any elements with that class?

**Reply by ju5t** [reply]
I should explain it better.

Counting elements with `wire:dirty` is already possible. Livewire registers them in a property called `dirtyEls`. What I'd like to know is how many of them are actually dirty; so I can show a popup before a user destroys all of the changes made. This is probably tracked somewhere, but I couldn't find it.

**Reply by joshhanley** [reply]
@ju5t no I understand 🙂

Yeah what I was suggesting would enable that.

So if you use a custom class, with `wire:dirty="is-dirty"` applied to all the inputs you want to track, then you could get all the `dirtyEls` and loop through and find any that actually contain the `is-dirty` class, as that will only be present on the element when it is actually dirty. 

Once you have that list, then you can do whatever you like with it.

Hopefully that makes sense!

**Reply by ju5t** [reply]
Oh, yea! That does make sense actually! I should have read your previous reply a few times over, sorry. I will give it a go. Thanks again!

**Reply by ju5t** [reply]
Hm, turns out I should have read the `wire:dirty` documentation a few times more too.

> There are cases where it may be useful to provide feedback that content has changed and is not yet in-sync with the back-end Livewire component.

So, setting `wire:dirty` only makes sense if you combine it with `wire:model.lazy` or `defer`. We sometimes do that, but not all the time. There are some edge cases.

I think what I will do is add `wire:dirty.class="is-dirty"` to each input element whenever possible + have the `updated()` method in Livewire set the `$isDirty` public property to true and have a small JS function that checks the current state of things.
